sql/core/src/test/resources/sql-tests/inputs/pgSQL/aggregates_part2.sql

```diff
@@ -7,207 +7,211 @@
+-- explain (verbose, costs off)
+-- select array(select sum(x+y) s
+--             from generate_series(1,3) y group by y order by s)
+--   from generate_series(1,3) x;
+-- select array(select sum(x+y) s
+--             from generate_series(1,3) y group by y order by s)
+--   from generate_series(1,3) x;
+
+-- [SPARK-27879] Implement bitwise integer aggregates(BIT_AND and BIT_OR)
 --
 -- test for bitwise integer aggregates
 --
-CREATE TEMPORARY TABLE bitwise_test(
-  i2 INT2,
-  i4 INT4,
-  i8 INT8,
-  i INTEGER,
-  x INT2,
-  y BIT(4)
-);
+-- CREATE TEMPORARY TABLE bitwise_test(
+--   i2 INT2,
+--   i4 INT4,
+--   i8 INT8,
+--   i INTEGER,
+--   x INT2,
+--   y BIT(4)
+-- );
 
 -- empty case
-SELECT
-  BIT_AND(i2) AS "?",
-  BIT_OR(i4)  AS "?"
-FROM bitwise_test;
-
-COPY bitwise_test FROM STDIN NULL 'null';
-1	1	1	1	1	B0101
-3	3	3	null	2	B0100
-7	7	7	3	4	B1100
-\.
-
-SELECT
-  BIT_AND(i2) AS "1",
-  BIT_AND(i4) AS "1",
-  BIT_AND(i8) AS "1",
-  BIT_AND(i)  AS "?",
-  BIT_AND(x)  AS "0",
-  BIT_AND(y)  AS "0100",
-
-  BIT_OR(i2)  AS "7",
-  BIT_OR(i4)  AS "7",
-  BIT_OR(i8)  AS "7",
-  BIT_OR(i)   AS "?",
-  BIT_OR(x)   AS "7",
-  BIT_OR(y)   AS "1101"
-FROM bitwise_test;
+-- SELECT
+--   BIT_AND(i2) AS "?",
+--   BIT_OR(i4)  AS "?"
+-- FROM bitwise_test;
+
+-- COPY bitwise_test FROM STDIN NULL 'null';
+-- 1	1	1	1	1	B0101
+-- 3	3	3	null	2	B0100
+-- 7	7	7	3	4	B1100
+-- \.
+
+-- SELECT
+--   BIT_AND(i2) AS "1",
+--   BIT_AND(i4) AS "1",
+--   BIT_AND(i8) AS "1",
+--   BIT_AND(i)  AS "?",
+--   BIT_AND(x)  AS "0",
+--   BIT_AND(y)  AS "0100",
+--
+--   BIT_OR(i2)  AS "7",
+--   BIT_OR(i4)  AS "7",
+--   BIT_OR(i8)  AS "7",
+--   BIT_OR(i)   AS "?",
+--   BIT_OR(x)   AS "7",
+--   BIT_OR(y)   AS "1101"
+-- FROM bitwise_test;
 
 --
 -- test boolean aggregates
 --
 -- first test all possible transition and final states
 
-SELECT
-  -- boolean and transitions
-  -- null because strict
-  booland_statefunc(NULL, NULL)  IS NULL AS "t",
-  booland_statefunc(TRUE, NULL)  IS NULL AS "t",
-  booland_statefunc(FALSE, NULL) IS NULL AS "t",
-  booland_statefunc(NULL, TRUE)  IS NULL AS "t",
-  booland_statefunc(NULL, FALSE) IS NULL AS "t",
-  -- and actual computations
-  booland_statefunc(TRUE, TRUE) AS "t",
-  NOT booland_statefunc(TRUE, FALSE) AS "t",
-  NOT booland_statefunc(FALSE, TRUE) AS "t",
-  NOT booland_statefunc(FALSE, FALSE) AS "t";
-
-SELECT
-  -- boolean or transitions
-  -- null because strict
-  boolor_statefunc(NULL, NULL)  IS NULL AS "t",
-  boolor_statefunc(TRUE, NULL)  IS NULL AS "t",
-  boolor_statefunc(FALSE, NULL) IS NULL AS "t",
-  boolor_statefunc(NULL, TRUE)  IS NULL AS "t",
-  boolor_statefunc(NULL, FALSE) IS NULL AS "t",
-  -- actual computations
-  boolor_statefunc(TRUE, TRUE) AS "t",
-  boolor_statefunc(TRUE, FALSE) AS "t",
-  boolor_statefunc(FALSE, TRUE) AS "t",
-  NOT boolor_statefunc(FALSE, FALSE) AS "t";
-
-CREATE TEMPORARY TABLE bool_test(
-  b1 BOOL,
-  b2 BOOL,
-  b3 BOOL,
-  b4 BOOL);
+-- SELECT
```

Comment:
Do we need to deal with `booland_statefunc` and `boolor_statefunc`? These 2 functions are not exist in doc: https://www.postgresql.org/docs/12/functions-aggregate.html

Reply:
```
  288  * Function for standard EVERY aggregate conforming to SQL 2003.
  289  * The aggregate is also named bool_and for consistency.
  290  *
  291  * Note: this is only used in plain aggregate mode, not moving-aggregate mode.
```

See: https://doxygen.postgresql.org/bool_8c_source.html

Reply:
`BooleanType and BooleanType` is equivalent to `bool bool_and bool`?

Reply:
Yes. it's equivalent. BOOL_AND and booland_statefunc, BOOL_OR and boolor_statefunc.

Reply:
Can we rewrite the queries by using AND on the columns with boolean data types?

Reply:
Do you mean:
```sql
select NULL and NULL is null AS "t"
```
?

Reply:
yes.

Reply:
The behaviour different:

1. Spark SQL:
```sql
SELECT
  -- boolean and transitions
  -- null because strict
  NULL and NULL IS NULL AS `t`,
  TRUE and NULL IS NULL AS `t`,
  FALSE and NULL IS NULL AS `t`,
  NULL and TRUE IS NULL AS `t`,
  NULL and FALSE IS NULL AS `t`,
  -- and actual computations
  TRUE and TRUE AS `t`,
  NOT TRUE and FALSE AS `t`,
  NOT FALSE and TRUE AS `t`,
  NOT FALSE and FALSE AS `t`
-- !query 1 schema
struct<t:boolean,t:boolean,t:boolean,t:boolean,t:boolean,t:boolean,t:boolean,t:boolean,t:boolean>
-- !query 1 output
NULL	true	false	false	false	true	false	true	false
``` 

PostgreSQL:
```sql
SELECT
  -- boolean and transitions
  -- null because strict
  booland_statefunc(NULL, NULL)  IS NULL AS "t",
  booland_statefunc(TRUE, NULL)  IS NULL AS "t",
  booland_statefunc(FALSE, NULL) IS NULL AS "t",
  booland_statefunc(NULL, TRUE)  IS NULL AS "t",
  booland_statefunc(NULL, FALSE) IS NULL AS "t",
  -- and actual computations
  booland_statefunc(TRUE, TRUE) AS "t",
  NOT booland_statefunc(TRUE, FALSE) AS "t",
  NOT booland_statefunc(FALSE, TRUE) AS "t",
  NOT booland_statefunc(FALSE, FALSE) AS "t";
 t | t | t | t | t | t | t | t | t 
---+---+---+---+---+---+---+---+---
 t | t | t | t | t | t | t | t | t
(1 row)
```

Reply:
You need to add parenthesis. 

```
%sql
SELECT
  -- boolean and transitions
  -- null because strict
  (NULL and NULL) IS NULL AS `t`,
  (TRUE and NULL) IS NULL AS `t`,
  (FALSE and NULL) IS NULL AS `t`,
  (NULL and TRUE) IS NULL AS `t`,
  (NULL and FALSE) IS NULL AS `t`,
  -- and actual computations
  TRUE and TRUE AS `t`,
  NOT (TRUE and FALSE) AS `t`,
  NOT (FALSE and TRUE) AS `t`,
  NOT (FALSE and FALSE) AS `t`
```

You will still see the difference. That is because of the strict mode.

Based on the doc of postgreSQL https://www.postgresql.org/docs/9.5/sql-createfunction.html
, for strict mode, 
> STRICT indicates that the function always returns null whenever any of its arguments are null. If this parameter is specified, the function is not executed when there are null arguments; instead a null result is assumed automatically.

Our AND does not have a strict mode. It is fine to have a behavior difference, but you need to update the comment to explain it.

Reply:
Thank you, I see.